recipes/generate.py

```diff
@@ -81,9 +81,40 @@ def _setup_model(
 
         return model
 
+    def convert_prompt_to_tokens(
+        self,
+        prompt: Union[DictConfig, str],
+        chat_format: ChatFormat,
+        instruct_template: InstructTemplate,
+    ) -> List[Message]:
+        # Should only be chat-style prompt or instruct-style prompt
+        if chat_format and instruct_template:
+            raise ValueError(
+                "Cannot pass both chat format and instruct template for generation"
+            )
+
+        # A single string can be formatted with instruct_template
+        if isinstance(prompt, str):
+            if instruct_template:
+                prompt = template.format(prompt)
+            return self._tokenizer.encode(prompt, add_bos=True, add_eos=False)
+
+        # Otherwise pass a DictConfig mapping role -> message, format with ChatFormat
+        # dict.items() will respect order for Python >= 3.7
+        else:
+            messages = [Message(role=k, content=v) for k, v in prompt.items()]
+            messages += [
+                Message(role="assistant", content="")
+            ]  # TODO: pass via config?
```

Comment:
why would you want to pass this via config?

Reply:
What if e.g. you don't want to explicitly add assistant in this format?

Reply:
but the prompt field in the config implies that it is system + user only message, it may be weird to include an assistant message even if it's just formatting

Reply:
I guess my comment is more around the assumption that we always want the tokenizer to tokenize a message with role="assistant" and empty body as the last part of the prompt. What if e.g. the user has a different role name, or wants to add "\n\n", or something like that?